_posts/18-05-2022-a_story_unfolds.md

```diff
@@ -0,0 +1,223 @@
+---
+layout: post
+title: A story unfolds
+category: Haskell
+description: or how I tried to plant a tree
+---
+# A story unfolds
+
+### or how I tried to plant a tree
+
+I tried to re-implement the Unix `tree` tool in Haskell, as a learning exercise. The goal is to create a tool that, when provided with a directory, lists all the contents and prints them in a way similar to this:
+```
+$ tree
+.
+├── a
+│   └── c
+└── b
+    ├── d
+    └── e
+
+5 directories, 0 files
+```
+It is well-defined, well-structured, and obviously recursive problem, so it should be easy to do, a no-brainer almost[^excuses].
+
+I started hacking away at the tree.
+
+The thing I would need first was a function to list directory contents, and another one to let me distinguish between directories and files. Quite fortunately, I had a pair of trusty axes on hand[^axes]:
+```haskell
+listDir :: File -> IO [File]
+isDir :: File -> IO Boolean
+```
+A naïve implementation came to mind right away[^madeup]:
+```haskell    
+tree :: File -> IO ()
+tree file = tree' "" file
+  where tree' prefix file = do
+          putStrLn (prefix ++ "- " ++ name file)
+          isDirectory <- isDir file
+          if isDirectory
+          then do
+            contents <- listDir file
+            traverse_ (tree' (' ':prefix)) contents
+          else return ()
+```
+It does not produce nice branch-looking lines between filenames, relying on indentation instead - but it is close enough.
+
+Obviously, this code is clumsy - and not just because the utility functions are crude. No, this code is just clumsy and inflexible in itself - and what's even worse, it makes one cardinal sin.
+_It mixes the responsibilities._
+
+While it is not the gravest sin of them all, as one could argue that _traversing_ the filesystem and _printing_ the filenames are very strongly related, it would be nice to do these two separately. In fact, the added flexibility might come in handy if we want to do something more than just printing - and, as we're about to see, at pretty much no added cost.
+
+So into the forest we go, preparing our tree-like data structure:
+```haskell
+data FileTree = RegularFile File | Directory File [FileTree]
+```
+Now we can write a function that grows these trees:
+```haskell
+buildTree :: File -> IO FileTree
+buildTree file = do
+  isDirectory <- isDir file
+  if isDirectory
+  then do
+    contents <- listDir file
+    subtrees <- traverse buildTree contents
+    return (Directory file subtrees)
+  else return (RegularFile file)
+```
+And we can restore the printing capability by writing another simple function:
+```haskell
+printTree :: FileTree -> String
+printTree = unlines . printTree'
+  where printTree' (RegularFile file) = ['-':(name file)]
+        printTree' (Directory file contents) = ('-':(name file)):(subtrees contents)
+        subtrees = map prepend . concatMap printTree'
```

Comment:
Huh parens would be useful here, it's not clear for me how this line is parsed :goncern:

Reply:
`(map prepend) . (concatMap printTree)'

Will fix that somehow.

Reply:
Ok... so in Scala terms it'd be:
```scala
        subtrees = (_.map(prepend)).compose(_.concatMap(printTree))'
```
where concatMap itself is `_.map(...).reduce(_ ++ _)`, I suppose?

Reply:
Yes. In this case it's even roughly equivalent to Scala's `flatMap`.